src/intro-to-c/debugging/gdb-intro.md

```diff
@@ -0,0 +1,369 @@
+---
+title: Introduction to GDB
+author: Johnson Laguerre
+---
+
+GDB is a very useful debugging tool that allows you to see what the values in your program look like at any given point in time, and it lets you temporarily modify those values to see how your program runs, acting as if you actually made those changes in your code.
+
+In the example below, a student is working on a program but encounters a segmentation fault (segfault). They're not sure what's causing it, but thanks to GDB, they can see what's going on "under the hood" of their program and try to solve the problem.
+
+## Case Study: A Pointer Program
+
+``` c
+#include <stdio.h>
+
+void multiply_by_two(int *current, int *arr);
+
+int main(void)
+{
+    int arr[5] = {10, 20, 30, 40, 50};
+    int *current = arr;
+    int i = 0;
+
+    printf("Before\n");
+    printf("------\n");
+
+    for (i = 0; i < 5; i++)
+    {
+        printf("arr[%d] = %d\n", i, *(current++));
+    }
+	
+    current = NULL;
+    multiply_by_two(current, arr);
+
+    printf("\nAfter\n");
+    printf("-----\n");
+
+    for (i = 0; i < 5; i++)
+    {
+        printf("arr[%d] = %d\n", i, *(current++));
+    }
+
+    printf("\nAll done!\n");
+
+    return 0;
+}
+
+void multiply_by_two(int *current, int *arr)
+{
+    int i = 0;
+    current = arr;
+
+    for (i = 0; i < 5; i++)
+    {
+        *(current++) = (*current * 2);
+    }
+	
+    current = arr;
+}
+
+```
+
+Okay, everything looks alright. Time to compile.
+
+```
+my-pc:~intro-to-c/assignments/pointers$ gcc my_file.c
+my-pc:~intro-to-c/assignments/pointers$
+```
+
+Yes, no warnings! Okay, now let me run it.
+
+```
+my-pc:~intro-to-c/assignments/pointers$ ./a.out
+Before
+------
+arr[0] = 10
+arr[1] = 20
+arr[2] = 30
+arr[3] = 40
+arr[4] = 50
+
+After
+-----
+Segmentation fault
+```
+
+No, not a segfault! This can't be possible--my code is *perfect*! Fine, I guess I'll just have to debug it. Time to start typing up print statements... or maybe I can try that debugger I've been reading about. What is it called again? Oh, yeah, GDB.
+
+## Compiling for Debugging
+
+Alright, first I have to recompile my code with the `-g` flag.
+
+```
+my-pc:~intro-to-c/assignments/pointers$ gcc -g my_file.c
```

Comment:
Next bonus tidbit: using `-g` will add debugging symbols, but using `-ggdb` will add alllll the debugging symbols *specifically with GDB in mind.*

From the documentation:
```
       -ggdb
           Produce debugging information for use by GDB.  This means to use the most expressive format available
           (DWARF, stabs, or the native format if neither of those are supported), including GDB extensions if at all
           possible.
```